File: RentalModule.Application/CommandHandlers/CarOrder/UpdateCarOrder/UpdateCarOrderCommandHandler.cs
using RentalModule.Application.Contract.CarOrders.UpdateCarOrder;
using RentalModule.Domain.Models;
using RentalModule.Domain.RepositoryInterfaces.CarOrder;
using Results.Application;
using Results.Domain;
using Shared.CQRS;
using Shared.CQRS.CommandHandlers;

namespace RentalModule.Application.CommandHandlers.CarOrder.UpdateCarOrder;

public class UpdateCarOrderCommandHandler : ICommandHandler<UpdateCarOrderCommand, HandlerResult<UpdateCarOrderResponse, IErrorResult>>
    {
        private readonly ICarOrderCommandRepository _commandRepository;
        private readonly ICarOrderQueryRepository _queryRepository;
        
        public UpdateCarOrderCommandHandler(
            ICarOrderCommandRepository commandRepository, 
            ICarOrderQueryRepository queryRepository)
        {
            _commandRepository = commandRepository;
            _queryRepository = queryRepository;
        }
        
        public async Task<HandlerResult<UpdateCarOrderResponse, IErrorResult>> Handle(UpdateCarOrderCommand request, CancellationToken cancellationToken)
        {
            var carOrder =  await _queryRepository.GetByIdAsync(request.Id, cancellationToken);
            
            if (carOrder is null) return new EntityNotFoundErrorResult()
            {
                Title = "Cannot update car order",
                Message = $"Car order with ID {request.Id} does not exist in the database."
            };
            
            var carOrderToUpdate = new CarOrderModel()
            {
                Id = carOrder.Id,
                UserId = request.UserId,
                CarOfferId = request.CarOfferId,
                StartDate = request.StartDate,
                EndDate = request.EndDate,
                Notes = request.Notes ?? carOrder.Notes,
                NumOfDrivers = request.NumOfDrivers,
                TotalCost = request.TotalCost
            };
            
            var updatedCarOrder = await _commandRepository.UpdateAsync(carOrderToUpdate, cancellationToken);
            
            return new UpdateCarOrderResponse()
            {
                Title = "Car order updated",
                Message = $"Car order with ID {updatedCarOrder.Id} was updated in the database."
            };
        }
    }

File: RentalApi/Application/command_handlers/car_order/update_car_order/update_car_order_command_handler.go
package commands

import (
	"context"
	contract "rental-api/Application.contract/car_orders/UpdateCarOrder"
	"rental-api/Application/services"
	models "rental-api/Domain/models/domestic"
	car_offer_repository "rental-api/Domain/repository_interfaces/car_offer_repository"
	repository_interfaces "rental-api/Domain/repository_interfaces/car_order_repository"
	"rental-api/Domain/responses"
	data_fetcher "rental-api/Domain/service_interfaces"
)

type UpdateCarOrderCommandHandler struct {
	carOrderCommandRepository repository_interfaces.CarOrderCommandRepository
	carOrderQueryRepository   repository_interfaces.CarOrderQueryRepository
	carOfferQueryRepository   car_offer_repository.CarOfferQueryRepository
	dataFetcher               data_fetcher.MicroserviceConnector
}

func NewUpdateCarOrderCommandHandler(
	carOrderCommandRepository repository_interfaces.CarOrderCommandRepository,
	carOrderQueryRepository repository_interfaces.CarOrderQueryRepository,
	carOfferQueryRepository car_offer_repository.CarOfferQueryRepository,
	dataFetcher data_fetcher.MicroserviceConnector,
) *UpdateCarOrderCommandHandler {
	return &UpdateCarOrderCommandHandler{
		carOrderCommandRepository: carOrderCommandRepository,
		carOrderQueryRepository:   carOrderQueryRepository,
		carOfferQueryRepository:   carOfferQueryRepository,
		dataFetcher:               dataFetcher,
	}
}

func (h *UpdateCarOrderCommandHandler) Handle(ctx context.Context, command *UpdateCarOrderCommand) (*contract.UpdateCarOrderResponse, error) {
	userInfo, err := h.dataFetcher.GetUserInternalInfo(command.JwtToken)
	if err != nil {
		response := responses.NewResponse[contract.UpdateCarOrderResponse](401, "Unauthorized")
		return &response, nil
	}

	existingOrder, err := h.carOrderQueryRepository.GetCarOrderByID(command.Id)
	if err != nil || existingOrder == nil {
		response := responses.NewResponse[contract.UpdateCarOrderResponse](404, "Car order not found")
		return &response, nil
	}

	if existingOrder.UserId != userInfo.ID && !services.IsAdminOrSuperAdmin(userInfo.Roles) {
		response := responses.NewResponse[contract.UpdateCarOrderResponse](403, "Not authorized to update this car order")
		return &response, nil
	}

	carOffer, err := h.carOfferQueryRepository.GetCarOfferByID(command.CarOfferId)
	if err != nil || carOffer == nil {
		response := responses.NewResponse[contract.UpdateCarOrderResponse](404, "Car offer not found")
		return &response, nil
	}

	carOrder := &models.CarOrderModel{
		Id:               command.Id,
		UserId:           existingOrder.UserId,
		CarOfferId:       command.CarOfferId,
		StartDate:        command.StartDate,
		EndDate:          command.EndDate,
		DeliveryLocation: command.DeliveryLocation,
		ReturnLocation:   command.ReturnLocation,
		NumOfDrivers:     command.NumOfDrivers,
		TotalCost:        command.TotalCost,
	}

	err = h.carOrderCommandRepository.UpdateCarOrder(ctx, carOrder)
	if err != nil {
		response := responses.NewResponse[contract.UpdateCarOrderResponse](500, "Failed to update car order")
		return &response, nil
	}

	return &contract.UpdateCarOrderResponse{
		BaseResponse: responses.NewBaseResponse(200, "Car order updated successfully"),
	}, nil
}
